keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1201 @@
+
+```yaml
+type: string
+format: dns1123subdomain
+```
+
+Kubernetes does not recognize this as a format. Today, Kubernetes treats
+[unrecognized formats as a no-op](https://github.com/kubernetes/kubernetes/blob/f9f9e7177a328ae47bac6b823d57a99dfbfd310b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/formats.go#L53-L65) and ignores them. So users would be able
+to write fields not conforming to this format to custom resources.
+
+If Kubernetes wishes to support this format, users workflows will break upon
+upgrade until the offending resources and identified and repaired.
+
+With this CRD Validation Ratcheting feature, Kubernetes would be able to
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation. But downgrading from an update that widens validation can be understood 
+semantically as an update that tightens validating.
+
+For example if they had unreletated issues with the new Kubernetes update and 
+wanted to downgrade:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) is implemented then we should expect it to also affect native type validation.
+
+#### Change in API Behavior
+
+This KEP changes assumptions that calls to Update will exhaustively validate
+every field of the CR. After this change, authors of CRs will need to do this test 
+with a CREATE rather than an UPDATE.
+
+NOTE: Errors that would have normally been thrown are converted into warnings,
+so authors using UPDATE to validate their CR still get feedback.
+
+### Risks and Mitigations
```

Comment:
Missing the risk that we are too loose in ratcheting. We can start with the easy cases of ratcheting and add more ratcheting. We cannot go back. That's a risk. Mitigation is to be conservative in adding ratcheting we don't fully understand, basically your "Some CRD Validations Are Not Ratchetable" section.

Reply:
Thanks for mentioning this, will add that

Reply:
Rephrased `Some CRD Validations Are Not Ratchetable` section to be the `Not All Rules Can Be Correctly/Easily Ratcheted`